### PR TITLE
services/tree: Generate timestamps based on node position in the container

### LINKED ISSUE #1442: Generate timestamp in log based on node position in the container
Different nodes must have generate timestamps to allow non-conflicting concurrent operations.
For received replicated operations, each timestamp should be validated

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1442, #1446, #1427 .
Depends on #1439 .

### CODE REVIEW

File: pkg/local_object_storage/pilorama/interface.go [hunk 1]
```diff
@@ -42,3 +46,19 @@ const (
 	AttributeFilename = "FileName"
 	AttributeVersion  = "Version"
 )
+
+// CIDDescriptor contains container ID and information about the node position
+// in the list of container nodes.
+type CIDDescriptor struct {
+	CID      *cidSDK.ID
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/pilorama/interface.go:53]: Yes, but let's do this after the SDK update.

File: pkg/services/tree/service.go [hunk 2]
```diff
@@ -470,3 +485,20 @@ func metaToProto(arr []pilorama.KeyValue) []*KeyValue {
 	}
 	return meta
 }
+
+var errNotInContainer = errors.New("node doesn't belong to a container")
+
+func (s *Service) getContainerInfo(cid *cidSDK.ID, pub []byte) (netmap.Nodes, int, int, error) {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/services/tree/service.go:496]: Fixed.

File: pkg/services/tree/cache.go [hunk 3]
```diff
@@ -15,7 +16,10 @@ type clientCache struct {
 	simplelru.LRU
 }
 
-const defaultClientCacheSize = 10
+const (
+	defaultClientCacheSize      = 10
+	defaultClientConnectTimeout = time.Second * 2
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/services/tree/cache.go:21]: Seems ok to me, given that all testing is done in local networks. This value will be made configurable.

File: pkg/services/tree/replicator.go [hunk 4]
```diff
@@ -19,22 +19,69 @@ type movePair struct {
...
-	defaultReplicatorTimeout  = time.Second * 2
+	defaultReplicatorCapacity    = 64
+	defaultReplicatorWorkerCount = 32
+	defaultReplicatorSendTimeout = time.Second * 5
 )
 
+func (s *Service) replicationWorker() {
+	for {
+		select {
+		case <-s.closeCh:
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/services/tree/replicator.go:36]: Fixed

**Review by carpawell — APPROVED** [review]

**Pull request merged**